awscli/clidriver.py

```diff
@@ -764,6 +779,8 @@ def __call__(self, args, parsed_globals):
                 # method of the operation caller. It represents the return
                 # code of the operation.
                 return override
+        elif getattr(parsed_args, 'subcommand', None) is not None:
```

Comment:
There's one implementation quirk that we need to work out here about arguments
and subcommands.  Suppose we have an alias file like this:

```
[command resourcegroupstaggingapi get-resources]
mysvc = --tag-filters Key=foo,Values=bar \
    Key=bar,Values=baz \
    --query ResourceTagMappingList[].[ResourceARN] --output text
```

The way it works now is if you run the command
`aws resourcegroupstaggingapi get-resources mysvc`, we'll see that there's
an alias defined for `mysvc`, do the text replacement, and the full command
expansion is:

```
aws resourcegroupstaggingapi get-resource \
    # This part below is from the "mysvc" alias definition.
    --tag-filters Key=foo,Values=bar Key=bar,Values=baz \
    --query ResourceTagMappingList[].[ResourceARN] --output text
```

So far so good.  The tricky part happens if we run a command like this:

```
aws resourcegroupstaggingapi get-resources mysvc --resource-type-filters ecs
```

The current behavior right now is that the `--resource-type-filters ecs` is
ignored.  No errors or anything, it's as if you didn't specify it at all.
On the surface that seems counterintuitive because I'd expect the final
expansion to merge with the alias definition and be:

```

aws resourcegroupstaggingapi get-resource \
    # This part is from the "mysvc" alias definition.
    --tag-filters Key=foo,Values=bar Key=bar,Values=baz \
    --query ResourceTagMappingList[].[ResourceARN] --output text \
    # This part from user's explicitly provided args
    --resource-type-filters ecs

```

However, this is not the current behavior of our
`awscli.customizations.commands.BasicCommand`, which I mirrored in the
`ServiceOperation` update.  Essentially, our convention has been that commands
will provide a subcommand table, or an arg table, but not both.  I checked that
this is the case for all of our existing customizations that subclass
`BasicCommand`.  This means that our logic when the command invoked is:

https://github.com/aws/aws-cli/blob/f23912d0b4e72c3b847d5fd74d63d101667f38d6/awscli/customizations/commands.py#L180-L190

In other words, if a subcommand was provided, we essentially throw away
any `parsed_args` and pass the subcommand the `remaining` (i.e. unknown)
args.  However if no subcommand is provided then we call `self._run_main`
with the parsed args.

So going back to the original example, when we invoke the command object for the
`mysvc` alias definition, we throw away the `--resource-type-filters ecs` part
because it was "handled" by the `ServiceOperation`.  It was part of the `parsed_args`
for the command object.  This hasn't mattered in the past because the only command
objects that provide an `arg_table` are the "leaf commands".  That is, by convention,
we don't support scoped args such as:

```
aws foo --local-1=a bar --local-2=b baz --local3=c
```

The expectation is that all of those args are handled by the `baz` subcommand.

I have a few ideas on how we can make this work for aliases (I think), but the question is:
what should the ideal behavior be for `aws resourcegroupstaggingapi get-resources mysvc --resource-type-filters ecs`?

Reply:
For the command:
```
aws resourcegroupstaggingapi get-resources mysvc --resource-type-filters ecs
```
I think we should respect the `--resource-type-filters ecs` arguments and not throw them away and instead merge it with the alias definition. It would definitely be confusing if we did not do this.

Reply:
In that case we have a few decisions to make.

The simplest way to address this is to first get the subcommand table, and if it's not empty, construct a parser for *just* the subcommand.  If you then find that `parsed_args.subcommand is not None`, then delegate to the subcommand and pass through the remaining args.  That way the `ServiceOperation` never consumes any known parsed args.

I can update the `ServiceOperation` to support this behavior because it never supported subcommands to begin with so there's no concern with backwards compatibility.  However, this would make the behavior inconsistent with the `awscli.customizations.BasicCommand`, from the code I linked above we can see that it will consume any parsed args you provide in its `ARG_TABLE` but still delegate to a subcommand if found.

This won't break any of the customizations that ship with the CLI, as they provide a subcommand table OR an arg table but never both, but this could potentially break existing commands that do this (though I can't imagine why they would as the args literally get thrown away if there's a subcommand).

Should I just update this for `ServiceOperation` for now?

Reply:
Would only supporting `ServiceOperation` mean we would not be supporting merging alias definitions with arguments provided for leaf commands subclassed from `BasicCommand`? If so, I'd definitely would want us to explore getting the behavior working for custom commands too as the custom commands are fairly popular and I'd imagine customers would want to make aliases for them as well (e.g. make an S3 filter alias if you have a long set of `--include`/`--exclude` filters you commonly use).

> this could potentially break existing commands that do this

Are you referring to external plugins here? I think if any of the builtin custom commands are safe from breakage, I'm fine with making the change as plugins are still provisional in v2. This change will likely go into a minor version bump anyway.

Reply:
> Would only supporting ServiceOperation mean we would not be supporting merging alias definitions with arguments provided for leaf commands subclassed from BasicCommand?

Yeah, I was hoping to minimize amount of files changed, but we'd need to update `BasicCommand` to support this which also makes this a riskier change.

> Are you referring to external plugins here? 

Yeah it would only affect external plugins.

I'll play around with different ideas, I'm starting to suspect this implementation might get a bit tricky.